lib/spack/spack/test/spec_semantics.py

```diff
@@ -1303,7 +1303,7 @@ def test_spec_trim(mock_packages, config):
 def test_concretize_partial_old_dag_hash_spec(mock_packages, config):
     # create an "old" spec with no package hash
     bottom = Spec("dt-diamond-bottom").concretized()
-    delattr(bottom, "_package_hash")
+    bottom._package_hash = None
```

Comment:
Was this creating an issue? Or is the change unrelated?

Reply:
Yes, if this change is removed, you get the failure in https://github.com/spack/spack/actions/runs/7894899814/job/21546544138

* This PR ends up hashing input specs as part of concretization (that wasn't done before)
  * Actually to be precise, it makes it so that specs are hashed via the two distinct concretization pathways we have (see https://github.com/spack/spack/pull/42655#discussion_r1488781515)
* This test provides an un-hashable input spec (`delattr` makes `set([spec])` fail)
* My first thought was that we should avoid hashing specs in general as part of this process (so I tried https://github.com/spack/spack/pull/42655#discussion_r1488810448)
* I then decided though that the specs hashed in this PR are concrete or user specs, which should always be OK to hash, and this test was violating the contract of the object, so I changed the simulation to match what would occur "naturally" if no package hash was available (the attribute should be present but `None`).

Reply:
I'll try to investigate this. I would have expected it to not fail, since:

https://github.com/spack/spack/blob/98de8e325708da62c0d17b4607e62c9c7b7ab9e2/lib/spack/spack/spec.py#L4288

So I wonder which code path we are hitting. Note that this is tricky, because we don't set this attribute in `Spec.__init__` but in a very indirect way using:

https://github.com/spack/spack/blob/98de8e325708da62c0d17b4607e62c9c7b7ab9e2/lib/spack/spack/spec.py#L1895

where the attribute that gets written is also computed indirectly:

https://github.com/spack/spack/blob/98de8e325708da62c0d17b4607e62c9c7b7ab9e2/lib/spack/spack/hash_types.py#L31-L34

TL;DR Given the above, I don't think using `None` is testing the same situation we'd have with old specs. There it's likely that the `_package_hash` doesn't exist as an attribute (but I need to double check to be 100% sure).

Reply:
I thought that because

```
delattr(spec, "_package_hash")
set([spec]) # <---- this fails
```

fails it meant that `delattr` was wrong, although after re-checking I think you're right that this should work (`Spec._finalize_concretization` claims some concrete Specs may not have this attr). I think the other approach is to update

```
def to_node_dict(...):
    if self._concrete and hash.package_hash and self._package_hash:
```

(i.e. the spot that generates the exception)

to

```
if self._concrete and hash.package_hash and hasattr(self, "_package_hash") and self._package_hash:
```

I'll say though that other methods of Spec may be inconsistent in what they expect of `_package_hash`, e.g. `Spec.package_hash` also seems to assume `_package_hash` would always be present.

Reply:
Not to be fixed in this PR, but as we are discussing it: are there any technical reason we use this convoluted method of setting a private attribute without declaring it in `__init__`?

Reply:
Not that I'm aware of. IMO setting `_package_hash = None`  in `__init__` would be better (but I could be overlooking logic that changes its behavior depending on whether the attr is present).